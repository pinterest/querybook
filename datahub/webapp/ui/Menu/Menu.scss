.Menu {
    background-color: var(--bg-color);
    color: var(--text-color);
    user-select: none;
    border-radius: 3px;

    .MenuItem,
    .MenuInfoItem {
        display: flex;
        flex-direction: row;
        align-items: center;
        position: relative;
        color: inherit;

        text-align: left;
        white-space: nowrap;
        width: 100%;
        padding: 8px 24px 8px 12px;
    }

    .MenuItem {
        cursor: pointer;

        .Menu-icon {
            width: 18px;
        }

        &:hover {
            background-color: var(--hover-bg-color);
            color: var(--hover-text-color);
            a {
                color: inherit;
            }
        }
        &:first-child {
            border-top-left-radius: 3px;
            border-top-right-radius: 3px;
        }
        &:last-child {
            border-bottom-left-radius: 3px;
            border-bottom-right-radius: 3px;
        }
    }

    &.soft {
        background: var(--light-bg-color);
        color: var(--light-text-color);
        border: none;

        .MenuItem:hover,
        .MenuItem.is-active {
            color: var(--text-color);
        }
    }
}
