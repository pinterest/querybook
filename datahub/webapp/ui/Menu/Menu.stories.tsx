import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import {
    Menu,
    MenuInfoItem,
    MenuItem,
    MenuItemPing,
    MenuDivider,
} from './Menu';

export const _Menu = () => (
    <Menu>
        <MenuItem>First Item</MenuItem>
        <MenuItem>
            Second Item with Ping
            <MenuItemPing />
        </MenuItem>
        <MenuDivider />
        <MenuInfoItem>Info Item</MenuInfoItem>
    </Menu>
);
export default {
    title: 'Stateless/Menu',
    decorators: [centered],
};
