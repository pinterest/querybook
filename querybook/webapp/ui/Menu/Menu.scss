.Menu {
    background-color: var(--bg);
    color: var(--text);
    user-select: none;
    border-radius: var(--border-radius-sm);
    // overflow: hidden;

    .MenuItem,
    .MenuInfoItem {
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: left;
        position: relative;
        color: inherit;

        text-align: left;
        white-space: nowrap;
        width: 100%;
        padding: 8px 24px 8px 12px;
    }

    .MenuItem {
        cursor: pointer;
        * {
            cursor: pointer;
        }

        &:hover {
            background-color: var(--bg-hover);
            color: var(--hover-text-color);
            a {
                color: inherit;
            }
        }
        &:first-child {
            border-top-left-radius: inherit;
            border-top-right-radius: inherit;
        }
        &:last-child {
            border-bottom-left-radius: inherit;
            border-bottom-right-radius: inherit;
        }
    }

    .MenuInfoItem {
        color: var(--text-light);
    }

    &.soft {
        background: var(--bg-light);
        color: var(--text-light);
        border: none;

        .MenuItem:hover,
        .MenuItem.is-active {
            color: var(--text);
        }
    }
}
