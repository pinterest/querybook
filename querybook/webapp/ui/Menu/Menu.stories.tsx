import centered from '@storybook/addon-centered/react';
import React from 'react';

import {
    Menu,
    MenuDivider,
    MenuInfoItem,
    MenuItem,
    MenuItemPing,
} from './Menu';

export const _Menu = () => (
    <Menu>
        <MenuItem>First Item</MenuItem>
        <MenuItem>
            Second Item with Ping
            <MenuItemPing />
        </MenuItem>
        <MenuDivider />
        <MenuInfoItem>Info Item</MenuInfoItem>
    </Menu>
);
export default {
    title: 'Stateless/Menu',
    decorators: [centered],
};
