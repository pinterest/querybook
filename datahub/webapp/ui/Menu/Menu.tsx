import styled from 'styled-components';
import './Menu.scss';

export const Menu = styled.div.attrs<{
    height: number;
}>({
    className: 'Menu',
})`
    ${(props) =>
        props.height &&
        `maxHeight: ${props.height};
        overflow-x: hidden;
        overflow-y: auto;`};
`;
export const MenuItem = styled.div.attrs<{}>({
    className: 'MenuItem',
})``;
export const MenuInfoItem = styled.div.attrs<{}>({
    className: 'MenuInfoItem',
})``;
export const MenuDivider = styled.div.attrs<{}>({
    className: 'MenuDivider',
})`
    background-color: var(--inner-border-color);
    display: block;
    height: 1px;
    margin: 0px;
`;
