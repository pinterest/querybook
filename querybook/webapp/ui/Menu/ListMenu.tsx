import clsx from 'clsx';
import React from 'react';

import { TooltipDirection } from 'const/tooltip';
import { Dropdown } from 'ui/Dropdown/Dropdown';
import { Icon } from 'ui/Icon/Icon';
import type { AllLucideIconNames } from 'ui/Icon/LucideIcons';
import { AccentText } from 'ui/StyledText/StyledText';

import { Menu, MenuItem } from './Menu';

export interface IListMenuItem {
    checked?: boolean;
    icon?: AllLucideIconNames;
    onClick?: () => any;
    link?: string;
    tooltip?: string;
    tooltipPos?: TooltipDirection;
    name: React.ReactChild;

    items?: IListMenuItem[];
}

interface IProps {
    items?: IListMenuItem[];
    type?: 'select'; // Optional, One of none or 'select'
    /**
     * Defines the nested menu behavior
     */
    isNestedRight?: boolean;
    height?: number;
    soft?: boolean;
    className?: string;
}

export const ListMenu: React.FunctionComponent<IProps> = ({
    items,
    type,
    isNestedRight,

    height,
    soft,
    className,
}) => {
    if (!items || !items.length) {
        return null;
    }

    const isSelect = type === 'select';

    const menuActionsDOM = items.map((action, index) => {
        const iconName = isSelect
            ? action.checked
                ? 'CheckCircle'
                : 'Circle'
            : action.icon
            ? action.icon
            : null;

        const actionProps: React.HTMLProps<HTMLAnchorElement> = {};
        if (action.onClick) {
            actionProps.onClick = action.onClick;
        }
        if (action.link) {
            actionProps.href = action.link;
        }
        if (action.tooltip) {
            actionProps['aria-label'] = action.tooltip;
            actionProps['data-balloon-pos'] = action.tooltipPos || 'left';
        }

        const buttonContent = (
            <span className="flex-row flex1">
                {iconName ? (
                    <Icon name={iconName} className="mr8" size={16} />
                ) : null}
                <span className="Menu-text flex1">{action.name}</span>
            </span>
        );
        let itemDOM: React.ReactChild;
        if (action.items) {
            itemDOM = (
                <Dropdown
                    className={clsx({
                        'Dropdown-nested-menu': true,
                        'nested-right': isNestedRight,
                        MenuItem: true,
                    })}
                    key={index}
                    customButtonRenderer={() => (
                        <a {...actionProps} className="flex-row">
                            {buttonContent}
                        </a>
                    )}
                    layout={[isNestedRight ? 'right' : 'left', 'top']}
                >
                    <ListMenu items={action.items} />
                </Dropdown>
            );
        } else {
            itemDOM = (
                <MenuItem {...actionProps} key={index}>
                    <AccentText>{buttonContent}</AccentText>
                </MenuItem>
            );
        }

        return itemDOM;
    });

    return (
        <Menu
            height={height}
            className={clsx({
                soft,
                [className]: className,
            })}
        >
            {menuActionsDOM}
        </Menu>
    );
};
