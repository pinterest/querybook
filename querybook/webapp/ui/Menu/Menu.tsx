import styled from 'styled-components';

import './Menu.scss';

export const Menu = styled.div.attrs<{
    height: number;
    boxShadow: boolean;
}>({
    className: 'Menu',
})`
    ${(props) =>
        props.height &&
        `maxHeight: ${props.height};
        overflow-x: hidden;
        overflow-y: auto;`}
    ${(props) => props.boxShadow && 'box-shadow: var(--box-shadow);'}
`;
export const MenuItem = styled.span.attrs({
    className: 'MenuItem',
})``;
export const MenuInfoItem = styled.div.attrs({
    className: 'MenuInfoItem',
})``;
export const MenuDivider = styled.div.attrs({
    className: 'MenuDivider',
})`
    display: block;
    height: 12px;
    margin: 0px;
`;
export const MenuItemPing = styled.div.attrs({
    className: 'MenuItemPing',
})`
    pointer-events: none;
    border-radius: 100px;
    background-color: var(--color-accent-light);
    width: 8px;
    height: 8px;
    margin-left: 8px;
    margin-right: -8px;
`;
