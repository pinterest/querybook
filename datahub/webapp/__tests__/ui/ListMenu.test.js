import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { ListMenu } from '../../ui/Menu/ListMenu';

const items = [{ name: 'test' }];

it('renders without crashing', () => {
    shallow(<ListMenu items={items} />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<ListMenu items={items} />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(<ListMenu items={items} />);
        expect(output).toMatchSnapshot();
    });
});
