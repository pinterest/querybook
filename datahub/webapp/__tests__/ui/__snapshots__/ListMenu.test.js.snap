// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<styled.div>
  <styled.div
    key="0"
  >
    <span
      className="flex-row"
    >
      <span
        className="Menu-text"
      >
        test
      </span>
    </span>
  </styled.div>
</styled.div>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<div
  className="sc-bdVaJa Menu kvKKjL"
>
  <div
    className="sc-bwzfXH MenuItem uTxCW"
  >
    <span
      className="flex-row"
    >
      <span
        className="Menu-text"
      >
        test
      </span>
    </span>
  </div>
</div>
`;
